README.rst

ILRT Django Oracle pool
=======================

Ed Crewe, `ILRT
<http://www.ilrt.bris.ac.uk/>`_ at University of Bristol, February 2010

Packaged version of http://code.djangoproject.com/ticket/7732 by Taras Halturin
django database backend that uses cx_Oracle session pooling for connections

Original Code Modifications
---------------------------

Pruned original ticket's base.py to just hold the pooling relevant code. 
Using the standard Oracle connection for the rest of the database classes, 
ie. operations, client and introspection. 

Tested with django 1.1 and 1.2, deals with the differences eg. to settings
handling. 

Extra features
--------------

- Added the pooling and logging parameters to the settings.

- The connector uses the standard python logging model and caters for logging 
  full details of queries, either to a file log or appending them to the 
  bottom of the screen if the log level is DEBUG.

- Added an option for running against existing (older) Oracle databases, ie those 
  which may not have unicode for character fields.

- Option also allows running the tests against an existing database so that 
  running tests doesnt require database creation (oracle sys dba) rights.

- Added a modification to the cursor to not parse parameters if not required.

Why use it?
-----------

Perhaps due to our remotely distributed Oracle network taking a very long time 
to establish connections, the use of cx_Oracle's session pooling for 
connections provided a truely radical performance boost for requests 
from 3-4 secs/req to 0.4 secs/req, so many times faster.   

For single direct Oracle access it might still give a doubling of performance. 
Install it and run the performance test to find out (see below).

Installation
------------

Download the egg (or use buildout) or download the tarball and extract it. 
Then add /path/to/django-oraclepool to your python path.

Specify DATABASE_ENGINE = 'oraclepool' instead of 'oracle' in your settings.  

If you dont want to use the default extra database settings then the following defaults
are used

>>> DEFAULT_EXTRAS = {'min':4,        # starting number of pooled connections
...                  'max':8,         # maximum number of connections in the pool
...                  'increment':1,   # increase by this amount when more are needed
...                  'homogeneous':1, # 1 = single credentials, 0 = multiple credentials
...                  'threaded':True, # server platform optimisation 
...                  'timeout':600,   # connection timeout, 600 = 10 mins
...                  'log':0,         # extra logging functionality turned on
...                  'logfile':'',    # file system path to log file
...                  'existing':''    # Type modifications for existing database and flag for tests
...                  'session':[]     # Add session optimisations applied to each fresh connection, eg.
...                                    #   alter session set cursor_sharing = similar;
...                                    #   Enables use of bind variables assuming it isnt set at a system level 
...				       #   alter session set session_cached_cursors = 20;
...                                    #   Allows cursor reuse between queries   
...                  }

Note that if you want sql logging to screen when in DEBUG mode then add 
'oraclepool.log_sql.SQLLogMiddleware' to your MIDDLEWARE_CLASSES


General Performance
-------------------

Initially after apache restart you will see the first few requests taking the same time as each 
one initiates a new pooled connection. Then request speed should drop as it loses the Oracle 
connection time.

NB: Note that if you have PythonDebug On then the pool may get flushed much more regularly. 
So you will often get the slower pool populating requests.

Using mod_wsgi rather than mod_python may give a 25% added increase although this needs
confirmation on a production instance.

It should be remembered that there is also a very great deal of performance work that can be 
done at the database level. I have posted a page with some of the tips and tricks for 
improving database performance on my blog - http://python.blogs.ilrt.org/database-performance/

Pooling alternatives
--------------------

Pyora pool

Also tried out pyora pool see http://code.google.com/p/pyorapool
but found increase was only around 90% and also had 
issues with connection control and database edits failing.
This also requires the whole architecture of a separate remote procedure call daemon 
that holds the connection pool. Uneccesary here ... although useful
for pooling across different applications, or multiple servers.

ORM pools

Usually ORMs have a generic pooling capability, unfortuately djangos only has a beta one 
in development, unless you plugin another ORM, eg. http://www.sqlalchemy.org/ instead. 
However that does require code rewriting.

Having said that a generic ORM level pool is unlikely to perform as well as one at the 
database connector level, which in turn is going to be less fast than one within the 
database itself (see below).

Tests
-----

The tests are run via separate test apps in the django-oraclepool folder.
Some of these tests are derived from a set for http://code.google.com/p/django-mssql/

They also include the option to run the test suite against an existing database for users who dont
have full oracle dba rights on their test oracle servers. 

The key extra tests are performance timings for running the test suite via the pooled oracle
connection vs the standard one. Hopefully these timings should indicate whether using oraclepool is
of value when using django with your oracle server network. 

Run the tests by running django-oraclepool/tests/manage.py test 
Or run individual tests by supplying there name, eg. manage.py test performance

The performance test simulates a real environment by running up a number of connections
as would exist with a production web server (the Apache2 default is 2 processes * 64 threads)
whilst the test creates a maximum of only 32.
In practise I found the actual performance improvement significantly greater than that indicated 
by the doubling of speed that the multiple connections performance test gives. However that
may not be the case dependent on your production oracle and web server environment.








docs/HISTORY.rst
-*- restructuredtext -*-

Changelog
=========

0.6 - UNRELEASED
----------------

- Fix a bug with adding sql log to page when in DEBUG

0.5 - Fix logger bug and add session config (2010-02-19)
--------------------------------------------------------

- Fix a bug preventing the logger from being used
- Make sure session settings are only set on connection creation not cursor use
- Add a session list of sql lines for optionally setting session parameters
- Wrap connection signal in try except for older django compatibility.

  [Ed Crewe]

0.4 - Fix performance test (2010-01-29)
---------------------------------------

- Whoops, I realised the performance test was underplaying the improvement due to a bug,
  it now gives the same performance for single connections (as expected) and 300% 
  improvement for multiple ones.
- Added signal for connection 'creation' and fixed raise of unicode error

  [Ed Crewe]

0.3 - Initial release (2010-01-28)
----------------------------------

- Added test suite including performance test to demonstrate benefits
- Tested and made compatible with django 1.2 (as well as django 1.1)
- Cleaned up the code and factored out anything that wasnt specific to pooling,
  subclassing most components from the standard Oracle engine.
- Modified logging options so DEBUG switches on log to screen and EXTRAS['log'] 
  switches logging to file option separately.

  [Ed Crewe]

0.2 - Used for UoB student reps
-------------------------------

- Commented out debug print statements from DatabaseWrapper._cursor which broke install in wsgi
- Added DatabaseWrapper.__init__ like the oracle backend with client, creation and introspection
  borrowed from it, since creation is expected in django 1.1.1 code.
- Added default value of None for settings in DatabaseWrapper._cursor since this argument is not 
  expected in django 1.1.1 code.
- Added DATABASE_EXTRAS dictionary to set all the session parameters from django config.
- Added session pool settings report method, to check these.
- Added python logger option and sql_log middleware

  [Ed Crewe]

0.1 - Unreleased
----------------

* Initial release - packaging of code from http://code.djangoproject.com/ticket/7732
 
  [Taras Halturin, et al.]


tests/settings.py
# Django settings for testbackend project.
# from dbsettings import *

# Set up common credentials for testing
# oraclepool vs. oracle functionality & performance 
CREDENTIALS = {'USER' : 'SCOTT',      
               'NAME' : 'SCOTT',             
               'PASSWORD' : 'TIGER',     
               'HOST' : 'localhost', 
               'PORT' : '1521',            
               'OPTIONS' : {}
               }

# NB: Oracle already has OPTIONS dict so use EXTRAS for pooling and logging
EXTRAS = {'min':4,         # start number of connections
          'max':8,         # max number of connections
          'increment':2,   # increase by this amount when more are needed
          'homogeneous':1, # 1 = single credentials, 0 = multiple credentials
          'threaded':True, # server platform optimisation 
          'timeout':10,   # connection timeout, 600 = 10 mins
          'log':0,         # extra logging functionality
          'logpath':'.',    # file system path to log file
          'existing':'Unicode',   # Type modifications if using existing database data
          'session': ["alter session set session_cached_cursors = 8;",
                      "alter session set cursor_sharing = 'SIMILAR'"]
          }

# Use django 1.2 or later multi-db setting (see rewrite for pre 1.2 below ...)
DATABASES= { 'oraclepool':{'ENGINE' : 'oraclepool',
                           'EXTRAS' : EXTRAS
                          },
             'oracle':{'ENGINE' : 'oracle',
                       'EXTRAS' : {'existing':'Unicode'}
                       }
             }

for db in ['oraclepool','oracle']:
    for key in CREDENTIALS.keys():
        DATABASES[db][key] = CREDENTIALS[key] 

### Switch backend for standard tests
### Specify oracle or oraclepool for running the test suite against
DATABASES['default'] = DATABASES['oraclepool']

def get_settings_dict(db='default'):
    """ Make compatible with 1.2 or earlier database settings
        used by apitest and performance test
    """   
    try:
        from settings import DATABASES
        settings_dict = DATABASES.get(db,{})
    except:
        pass
    if not settings_dict:
        for key in ['HOST','PORT','NAME','USER','PASSWORD','ENGINE','OPTIONS','EXTRAS']:
            settings_dict[key] = globals().get('DATABASE_%s' % key, '') 
    for key in settings_dict.keys():
        settings_dict['DATABASE_%s' % key] = settings_dict[key]
    return settings_dict

# Also rewrite into globals for pre django 1.2 compatibility
for k, v in DATABASES['default'].items():
    globals()['DATABASE_' + k] = v

DEBUG = True
TEMPLATE_DEBUG = DEBUG

ADMINS = (
    # ('Your Name', 'your_email@domain.com'),
)

MANAGERS = ADMINS

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# If running in a Windows environment this must be set to the same as your
# system time zone.
TIME_ZONE = 'America/New York'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en-us'

SITE_ID = 1

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# Absolute path to the directory that holds media.
# Example: "/home/media/media.lawrence.com/"
MEDIA_ROOT = ''

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash if there is a path component (optional in other cases).
# Examples: "http://media.lawrence.com", "http://example.com/media/"
MEDIA_URL = ''

# URL prefix for admin media -- CSS, JavaScript and images. Make sure to use a
# trailing slash.
# Examples: "http://foo.com/media/", "/media/".
ADMIN_MEDIA_PREFIX = '/media/'

# Make this unique, and don't share it with anybody.
SECRET_KEY = '%(u8uds4xsy0+95cj3o&k49*u@&--yp0t&e&0$!@s2fvea#u4j'

# List of callables that know how to import templates from various sources.
TEMPLATE_LOADERS = (
    'django.template.loaders.filesystem.load_template_source',
    'django.template.loaders.app_directories.load_template_source',
)

MIDDLEWARE_CLASSES = (
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.middleware.doc.XViewMiddleware',
)

ROOT_URLCONF = 'urls'

TEMPLATE_DIRS = (
    # Put strings here, like "/home/html/django_templates" or "C:/www/django/templates".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
)

INSTALLED_APPS = (
    'performance',
    'apitest',
    'regressiontests',
    'slicing',
    'nulls',
    'aggregates'
)


oraclepool/log_sql.py
# Middleware added if log level on INFO or DEBUG
# If dango DEBUG then write sql timings to screen else write to the log 
from django.conf import settings 
from django.db import connection
from django.template import Template, Context
from django.utils.html import strip_tags
logger = None
try:
    from oraclepool.base import get_extras
    from datetime import datetime
    if int(get_extras().get('log', 0)) == 10:
        import logging
        logger = logging.getLogger("oracle_pool")
except:
    pass

TMPL = Template('''
                <p>Session pool config: {{ pool }}</p> 
                <p><em>Total query count:</em> {{ count }}<br/>
                <em>Total execution time:</em> {{ time }}</p>
                <ul class="sqllog">
                    {% for sql in sqllog %}
                        <li>{{ sql.time }}: {{ sql.sql }}</li>
                    {% endfor %}
                </ul>
                ''')

class SQLLogMiddleware:
    """ Writes the pool details and sql query timings either to the log
        or at bottom of the page if django DEBUG is set
    """

    def process_response ( self, request, response ):
        """ SQL logging to screen switched on by settings.DEBUG
            SQL logging to log dependent on database logger level
            ie. EXTRAS['log'] = 10
            Use case is need to turn on debug SQL logging on production installs
            occasionally - whilst DEBUG is only for development
        """
        if settings.DEBUG or logger:
            time = 0.0
            for q in connection.queries:
                time += float(q['time'])

            if settings.DEBUG:
                if response.__getitem__('Content-Type').startswith('text/html'):

                    content = TMPL.render(Context({'pool':self.pool(),
                                                   'sqllog':connection.queries,
                                                   'count':len(connection.queries),
                                                   'time':time}))

                if response.content.find('</body>') == -1:
                    response.content += content
                else:
                    response.content = response.content.replace('</body>',
                                                  "%s%s" % (content,'</body>'))

            if logger:
                logger.debug('%s SQL log for %s -----------------------' % (datetime.now(),
                                                                            request.path) )
                logger.debug(self.pool())
                logger.debug('Executed %s queries in %s' % (len(connection.queries),time))
                for query in connection.queries:
                    logger.debug(query)
        return response


    def pool(self):
        """ show pooling info """
        if hasattr(connection, 'get_config'):
            return str(connection.get_config())
        else:
            return 'No connection pooling implemented'


oraclepool/base.py
"""
Oracle pooled connection database backend for Django.
Requires cx_Oracle: http://www.python.net/crew/atuining/cx_Oracle/
"""

import os, sys
from django.db.backends import BaseDatabaseWrapper, BaseDatabaseValidation, util
try:
    from django.db.backends.signals import connection_created
except:
    connection_created = None
# Makes it explicit where the default oracle versions fo these components are used
from django.db.backends.oracle.base import OracleParam
from django.db.backends.oracle.base import DatabaseFeatures as OracleDatabaseFeatures
from django.db.backends.oracle.base import DatabaseOperations as OracleDatabaseOperations
from django.db.backends.oracle.client import DatabaseClient as OracleDatabaseClient
from django.db.backends.oracle.introspection import DatabaseIntrospection as OracleDatabaseIntrospection
from django.db.backends.oracle.base import FormatStylePlaceholderCursor as OracleFormatStylePlaceholderCursor

from oraclepool.creation import DatabaseCreation
from django.conf import settings

try:
    import cx_Oracle as Database
except ImportError, e:
    from django.core.exceptions import ImproperlyConfigured
    raise ImproperlyConfigured("Error loading cx_Oracle module: %s" % e)

from django.utils.encoding import smart_str, force_unicode
# Check whether cx_Oracle was compiled with the WITH_UNICODE option.  This will
# also be True in Python 3.0.
if int(Database.version.split('.', 1)[0]) >= 5 and not hasattr(Database, 'UNICODE'):
    convert_unicode = force_unicode
else:
    convert_unicode = smart_str

DatabaseError = Database.Error
IntegrityError = Database.IntegrityError

# Oracle takes client-side character set encoding from the environment.
os.environ['NLS_LANG'] = '.UTF8'

def get_extras(database='default'):
    """ NB: Oracle already has OPTIONS specific to cx_Oracle.connection() use
        This adds extra pool and sql logging attributes to the settings 
    """
    DEFAULT_EXTRAS = {'min':4,         # start number of connections
                      'max':8,         # max number of connections
                      'increment':1,   # increase by this amount when more are needed
                      'homogeneous':1, # 1 = single credentials, 0 = multiple credentials
                      'threaded':True, # server platform optimisation 
                      'timeout':600,   # connection timeout, 600 = 10 mins
                      'log':0,         # extra logging functionality
                      'logpath':'',    # file system path for oraclepool.log file
                      'existing':'',   # Type modifications if using existing database data
                      'session':[]     # Add session optimisations applied to each fresh connection, eg.
                                       #   ['alter session set cursor_sharing = similar',
                                       #    'alter session set session_cached_cursors = 20']
                      }
    if hasattr(settings, 'DATABASES'):
        db = settings.DATABASES.get(database,{})
        if db.has_key('EXTRAS'):
            return db['EXTRAS']
    elif hasattr(settings, 'DATABASE_EXTRAS'):
        return settings.DATABASE_EXTRAS
    else:
        return DEFAULT_EXTRAS

def get_logger(extras):
    """ Check whether logging is required
        If log level is more than zero then logging is performed
        If log level is DEBUG then logging is printed to screen
        If no logfile is specified then unless its DEBUG to screen its added here
        NB: Log levels are 10 DEBUG, 20 INFO, 30 WARNING, 40 ERROR, 50 CRITICAL
    """

    loglevel = int(extras.get('log', 0))
    if loglevel>0:
        import logging
        logfile = extras.get('logpath','')
        if logfile.endswith('.log'):
            (logfile, filename) = os.path.split(logfile)
        else:
            filename = 'oraclepool.log'
        if os.path.exists(logfile):
            logfile = os.path.join(logfile,filename)
        else:
            logfile = ''
        if not logfile and extras.get('log') > logging.DEBUG:
            logfile = '.'
        if logfile in ['.','..']:
            logfile = os.path.join(os.path.abspath(os.path.dirname(logfile)),filename)
        # if log file is writable do it
        if not logfile:
            print 'Log path %s not found' % extras.get('logpath','')
            return None
        else:
            try:
                logging.basicConfig(filename=logfile, level=loglevel)
                logger = logging.getLogger("oracle_pool")
                logger.setLevel(loglevel)
                ch = logging.StreamHandler()
                ch.setLevel(loglevel)
                formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
                ch.setFormatter(formatter)
                logger.addHandler(ch)
                from datetime import datetime
                msg = '%s #### Started django-oraclepool SQL logging at level %s ####' % (datetime.now(),
                                                                                             loglevel)
                print 'Log started at %s' % logfile
                logger.info(msg)
                return logger
            except:
                print 'Log configuration failed' 
                return None
    else:
        print 'No logging set so exceptions will be raised not logged'
        return None

    # Add sql logging for all requests if DEBUG level
    if extras.get('log') == 10 or settings.DEBUG:
        # Add middleware if needed
        MIDDLEWARE_CLASSES = list(settings.MIDDLEWARE_CLASSES) 
        MIDDLEWARE_CLASSES.append('oraclepool.log_sql.SQLLogMiddleware')
        settings.MIDDLEWARE_CLASSES = tuple(MIDDLEWARE_CLASSES)

DATABASE_EXTRAS = get_extras()
logger = get_logger(DATABASE_EXTRAS)
    
class DatabaseFeatures(OracleDatabaseFeatures):
    """ Add extra options from default Oracle ones
        Plus switch off save points and id return
        See http://groups.google.com/group/django-developers/browse_thread/thread/bca33ecf27ff5d63
        Savepoints could be turned on but are not needed
        and since they may impact performance they are turned off here 
    """
    uses_savepoints = False
    can_return_id_from_insert = False
    
    allows_group_by_ordinal = False
    supports_tablespaces = True
    uses_case_insensitive_names = True
    time_field_needs_date = True
    date_field_supports_time_value = False

class DatabasePool:
    """ Core connection pool """
    def __get__(self):
        if hasattr(self,'_pool'):
            return self._pool
        else:
            if len(settings.DATABASE_HOST.strip()) == 0:
                settings.DATABASE_HOST = 'localhost'
            if len(settings.DATABASE_PORT.strip()) != 0:
                dsn = Database.makedsn(settings.DATABASE_HOST, int(settings.DATABASE_PORT), settings.DATABASE_NAME)
            else:
                dsn = settings.DATABASE_NAME
            self._pool = cx_Oracle.SessionPool(settings.DATABASE_USER, 
                                     settings.DATABASE_PASSWORD, 
                                     dsn, 
                                     DATABASE_EXTRAS['min'], 
                                     DATABASE_EXTRAS['max'], 
                                     DATABASE_EXTRAS['increment'],
                                     threaded = DATABASE_EXTRAS['threaded'], 
                                     homogeneous = DATABASE_EXTRAS['homogeneous'])
            if DATABASE_EXTRAS['timeout']:
                self._pool.timeout = DATABASE_EXTRAS['timeout']
            return self._pool


class DatabaseWrapper(BaseDatabaseWrapper):
    """ This provides the core connection object wrapper
        for cx_Oracle's pool handling.
        The code is mostly taken from
        http://code.djangoproject.com/ticket/7732 by halturin
    """

    poolprops = {'homogeneous':'',
                 'increment':'',
                 'max':'',
                 'min':'',
                 'busy':'',
                 'opened':'',
                 'name':'',
                 'timeout':'',
                 'tnsentry':''
                }
    operators = {
        'exact': '= %s',
        'iexact': '= UPPER(%s)',
        'contains': "LIKEC %s ESCAPE '\\'",
        'icontains': "LIKEC UPPER(%s) ESCAPE '\\'",
        'gt': '> %s',
        'gte': '>= %s',
        'lt': '< %s',
        'lte': '<= %s',
        'startswith': "LIKEC %s ESCAPE '\\'",
        'endswith': "LIKEC %s ESCAPE '\\'",
        'istartswith': "LIKEC UPPER(%s) ESCAPE '\\'",
        'iendswith': "LIKEC UPPER(%s) ESCAPE '\\'",
    }
    oracle_version = None

    def __init__(self, *args, **kwargs):
        """ Set up the various database components
            Oracle prefixed classes use the standard Oracle
            version - tested with both 1.0 and 1.2
        """
        super(DatabaseWrapper, self).__init__(*args, **kwargs)

        self.features = DatabaseFeatures()
        self.ops = OracleDatabaseOperations()
        self.client = OracleDatabaseClient(self)
        self.creation = DatabaseCreation(self)
        self.introspection = OracleDatabaseIntrospection(self)
        try:
            self.validation = BaseDatabaseValidation(self)
        except:
            # pre django 1.2
            self.validation = BaseDatabaseValidation()            

    def get_config(self):
        """ Report the oracle connection and pool data see 
            http://cx-oracle.sourceforge.net/html/session_pool.html#sesspool
        """
        pool = self._get_pool()
        if pool:
            for key in self.poolprops.keys():
                try:
                    self.poolprops[key] = getattr(pool, key, '')
                except:
                    pass
        else:
            self.poolprops['name'] = 'Session pool not found'
        return self.poolprops

    def _get_pool (self):
        """ Get the connection pool or create it if it doesnt exist """
        if not hasattr (self.__class__, '_pool'):
            if DATABASE_EXTRAS['threaded']:
                Database.OPT_Threading = 1
            else:
                Database.OPT_Threading = 0
            if len(settings.DATABASE_HOST.strip()) == 0:
                settings.DATABASE_HOST = 'localhost'
            if len(settings.DATABASE_PORT.strip()) != 0:
                dsn = Database.makedsn(settings.DATABASE_HOST, int(settings.DATABASE_PORT), settings.DATABASE_NAME)
            else:
                dsn = settings.DATABASE_NAME
            try:
                p = Database.SessionPool(settings.DATABASE_USER, 
                                         settings.DATABASE_PASSWORD, 
                                         dsn, 
                                         DATABASE_EXTRAS['min'], 
                                         DATABASE_EXTRAS['max'], 
                                         DATABASE_EXTRAS['increment'],
                                         threaded = DATABASE_EXTRAS['threaded'], 
                                         homogeneous = DATABASE_EXTRAS['homogeneous'])
            except:
                print "##### Database '%s' login failed or database not found #####" % settings.DATABASE_NAME
                print "dsn = %s" % dsn
                print 'Django start up cancelled'
                sys.exit(1)
            if DATABASE_EXTRAS['timeout']:
                p.timeout = DATABASE_EXTRAS['timeout']
            setattr(self.__class__, '_pool', p)

        return getattr(self.__class__, '_pool')
        
    pool = property (_get_pool)

    def _cursor(self, settings=None):
        """ Get a cursor from the connection pool """
        cursor = None
        if self.pool is not None:
            if self.connection is None:
                # Set oracle date to ansi date format.  This only needs to execute
                # once when we create a new connection. 
                self.connection = self.pool.acquire()
                if connection_created:
                    # Assume acquisition of existing connection = create for django signal
                    connection_created.send(sender=self.__class__)
                if logger:
                    logger.info("Acquire pooled connection \n%s\n" % self.connection.dsn)

                cursor = FormatStylePlaceholderCursor(self.connection)
                cursor.execute("ALTER SESSION SET NLS_DATE_FORMAT = 'YYYY-MM-DD' "  
                               "NLS_TIMESTAMP_FORMAT = 'YYYY-MM-DD HH24:MI:SS.FF'")
                if DATABASE_EXTRAS.get('session',[]):
                    for sql in DATABASE_EXTRAS['session']:
                        cursor.execute(sql)
                try:
                    self.oracle_version = int(self.connection.version.split('.')[0])
                    # There's no way for the DatabaseOperations class to know the
                    # currently active Oracle version, so we do some setups here.
                    # TODO: Multi-db support will need a better solution (a way to
                    # communicate the current version).
                    if self.oracle_version <= 9:
                        self.ops.regex_lookup = self.ops.regex_lookup_9
                    else:
                        self.ops.regex_lookup = self.ops.regex_lookup_10
                except ValueError, err:
                    if logger:
                        logger.warn(str(err))
                try:
                    self.connection.stmtcachesize = 20
                except:
                    # Django docs specify cx_Oracle version 4.3.1 or higher, but
                    # stmtcachesize is available only in 4.3.2 and up.
                    pass
            else:
                cursor = FormatStylePlaceholderCursor(self.connection)                
        else:
            if logger:
                logger.critical('Pool couldnt be created')
            else:
                raise 'Pool couldnt be created'
            
        if not cursor:
            cursor = FormatStylePlaceholderCursor(self.connection)
        # Default arraysize of 1 is highly sub-optimal.
        cursor.arraysize = 100
        return cursor

    def close(self):
        """ Releases connection back to pool """
        if self.connection is not None:
            if logger:
                logger.debug("Release pooled connection\n%s\n" % self.connection.dsn)
            self.pool.release(self.connection)
            self.connection = None

    def _savepoint_commit(self, sid):
        """ Oracle doesn't support savepoint commits.  Ignore them. """
        pass

class FormatStylePlaceholderCursor(OracleFormatStylePlaceholderCursor):
    """ Added just to allow use of % for like queries without params
        and use of logger if present.
    """

    def cleanquery(self, query, args=None):
        """ cx_Oracle wants no trailing ';' for SQL statements.  For PL/SQL, it
            it does want a trailing ';' but not a trailing '/'.  However, these
            characters must be included in the original query in case the query
            is being passed to SQL*Plus.

            Split out this as a function and allowed for no args so
            % signs can be used in the query without requiring parameterization
        """
        if query.endswith(';') or query.endswith('/'):
            query = query[:-1]
        if not args:
            return convert_unicode(query, self.charset)
        else:
            try:
                return convert_unicode(query % tuple(args), self.charset)
            except TypeError, e:
                err = 'Parameter parsing failed due to error ' + str(e) + ' for query:' + query
                if logger:
                    logger.critical(err)
                else:
                    raise str(err)
                
    def execute(self, query, params=[]):
        if params is None:
            args = None
        else:
            params = self._format_params(params)
            args = [(':arg%d' % i) for i in range(len(params))]
        query = self.cleanquery(query, args)
        self._guess_input_sizes([params])
        try:
            return self.cursor.execute(query, self._param_generator(params))
        except DatabaseError, e:
            # cx_Oracle <= 4.4.0 wrongly raises a DatabaseError for ORA-01400.
            if e.args[0].code == 1400 and not isinstance(e, IntegrityError):
                e = IntegrityError(e.args[0])
            err = '%s due to query:%s' % (e, query)
            if logger:
                logger.critical(err)
            else:
                raise err

    def executemany(self, query, params=[]):
        try:
            args = [(':arg%d' % i) for i in range(len(params[0]))]
        except (IndexError, TypeError):
            # No params given, nothing to do
            return None
        query = self.cleanquery(query, args)        
        formatted = [self._format_params(i) for i in params]
        self._guess_input_sizes(formatted)
        try:
            return self.cursor.executemany(query,
                                [self._param_generator(p) for p in formatted])
        except DatabaseError, e:
            # cx_Oracle <= 4.4.0 wrongly raises a DatabaseError for ORA-01400.
            if e.args[0].code == 1400 and not isinstance(e, IntegrityError):
                e = IntegrityError(e.args[0])
            if logger:
                logger.critical('%s due to query:%s' % (e, query))                
            else:
                raise str(e)
